### PR TITLE
Automatically create stream if it doesn't exist.

### PR DESCRIPTION
**Pull request by benguillet**

### CODE REVIEW

File: drainer.go [hunk 1]
```diff
@@ -10,17 +14,38 @@ type Drainer struct {
...
-		return nil, err
+		logErr(err)
 	}
-
 	d := &Drainer{Buffer: buffer}
-	// Every second, we flush the buffer
-	go d.Drain()
 
-	return d, nil
+	fmt.Printf("\n\n\n\n\n\n\nSTREAM NAME : %s\n\n\n\n\n", streamName)
```
Comment (phobologic) [inline drainer.go]: Debug statement?

File: drainer.go [hunk 2]
```diff
@@ -10,17 +14,38 @@ type Drainer struct {
...
+	_, err = a.Client.CreateStream(&kinesis.CreateStreamInput{
+		ShardCount: aws.Int64(1),
+		StreamName: aws.String(streamName),
+	})
+
+	// If an error occur, return. Unless it's because the stream already exists.
+	if err != nil {
+		fmt.Println(err)
+		if reqErr, ok := err.(awserr.RequestFailure); ok {
+			fmt.Println(reqErr)
```
Comment (phobologic) [inline drainer.go]: This seems like it'd get noisy - and it'd probably look like an actual error.

File: drainer.go [hunk 3]
```diff
@@ -10,17 +14,38 @@ type Drainer struct {
...
+			if reqErr.Code() == "ResourceInUseException" {
+				a.addDrainer(streamName, d)
+			} else {
+				logErr(err)
+			}
+		} else {
+			logErr(err)
+		}
+	} else {
+		// Wait for the stream to be active.
```
Comment (phobologic) [inline drainer.go]: do we need addDrainer() here, since it won't be called if there is no error?

File: drainer.go [hunk 4]
```diff
@@ -29,3 +54,23 @@ func (d *Drainer) Drain() {
...
+func waitForActive(a *KinesisAdapter, d *Drainer) {
+	resp := &kinesis.DescribeStreamOutput{}
+	// timeout := make(chan bool, 30)
+	streamName := *d.Buffer.input.StreamName
+
+	params := &kinesis.DescribeStreamInput{StreamName: aws.String(streamName)}
+	for {
+		resp, _ = a.Client.DescribeStream(params)
+		if streamStatus := *resp.StreamDescription.StreamStatus; streamStatus == "ACTIVE" {
+			log.Printf("kinesis: STREAM '%s' STATUS: %s\n", streamName, streamStatus)
```
Comment (phobologic) [inline drainer.go]: Seems like you could move this out of this if statement:

```
streamStatus := *resp.StreamDescription.StreamStatus
log.Printf(...)
if streamStatus == "ACTIVE" {
...
} else {
...
}
```

File: utils.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+type missingEnvVarError struct {
+	envVar   string
+	tmplName string
+}
+
+func (e *missingEnvVarError) Error() string {
+	return fmt.Sprintf("The %s template is missing. Please set the %s env variable\n", e.tmplName, e.envVar)
+}
+
+func compileTmpl(envVar string, tmplName string) (*template.Template, error) {
```
Comment (phobologic) [inline utils.go]: Make any sense for this to have an optional default?

**Comment by benguillet** [comment]
@phobologic aahhh damnit I should'nt have open this PR, this is not ready for review ahah.

File: drainer.go [hunk 6]
```diff
@@ -10,17 +13,30 @@ type Drainer struct {
...
+				a.addDrainer(streamName, d)
+			} else {
+				logErr(err)
+			}
+		} else {
+			logErr(err)
+		}
+	} else {
+		log.Printf("kinesis: need to create stream for %s\n", streamName)
+		go waitForActive(a, d)
```
Comment (bmarini) [inline drainer.go]: If newDrainer is already running in a new goroutine, does this need to run in a separate goroutine?

File: drainer.go [hunk 7]
```diff
@@ -29,3 +45,31 @@ func (d *Drainer) Drain() {
...
+
+	params := &kinesis.DescribeStreamInput{StreamName: aws.String(streamName)}
+	resp := &kinesis.DescribeStreamOutput{}
+	for {
+		resp, _ = a.Client.DescribeStream(params)
+		if streamStatus = *resp.StreamDescription.StreamStatus; streamStatus == "ACTIVE" {
+			a.addDrainer(streamName, d)
+			break
+		} else {
+			time.Sleep(4 * time.Second)
```
Comment (bmarini) [inline drainer.go:70]: How long does it normally take?

Comment (benguillet) [inline drainer.go:70]: About 20-25 seconds

**Comment by bmarini** [comment]
Looks good to me. I imagine some people wouldn't be comfortable with a log adapter automatically creating new kinesis streams, so making it configurable might be nice. :+1:

**Comment by benguillet** [comment]
@bmarini will do in next pr

**Pull request merged**